crypto/ml_dsa/ml_dsa_encoders.c

```diff
@@ -0,0 +1,995 @@
+        *out = z2 >> 10;
+    } while (in < end);
+    return 1;
+}
+
+/*
+ * @brief Reverses the procedure of poly_encode_signed_two_to_power_17().
+ * See FIPS 204, Algorithm 19, BitUnpack(v, a, b) where a = 2^17 - 1, b = 2^17.
+ *
+ * @param p A polynomial to write coefficients to.
+ * @param pkt A packet object to read 576 encoded bytes from.
+ *
+ * @returns 1 on success, or 0 on error.
+ */
+static int poly_decode_signed_two_to_power_17(POLY *p, PACKET *pkt)
+{
+    uint32_t *out = p->coeff;
+    const uint32_t *end = out + ML_DSA_NUM_POLY_COEFFICIENTS;
+    const uint8_t *in;
+    static const uint32_t range = 1u << 17;
+    static const uint32_t mask_18_bits = (1u << 18) - 1;
+
+    do {
+        uint32_t a1, a2, a3;
+
+        if (!PACKET_get_bytes(pkt, &in, 9))
+            return 0;
+        in = OSSL_CRYPTO_load_u32_le(&a1, in);
+        in = OSSL_CRYPTO_load_u32_le(&a2, in);
+        a3 = (uint32_t) *in;
+
+        *out++ = mod_sub(range, a1 & mask_18_bits);
+        *out++ = mod_sub(range, (a1 >> 18) | ((a2 & 0xF) << 14));
+        *out++ = mod_sub(range, (a2 >> 4) & mask_18_bits);
+        *out++ = mod_sub(range, (a2 >> 22) | (a3 << 10));
+    } while (out < end);
+    return 1;
+}
+
+/*
+ * @brief Encode the public key as an array of bytes.
+ * See FIPS 204, Algorithm 22, pkEncode().
+ *
+ * @param key A key object containing public key values. The encoded public
+ *            key data is stored in this key.
+ * @returns 1 if the public key was encoded successfully or 0 otherwise.
+ */
+int ossl_ml_dsa_pk_encode(ML_DSA_KEY *key)
+{
+    int ret = 0;
+    size_t i, written = 0;
+    const POLY *t1 = key->t1.poly;
+    size_t t1_len = key->t1.num_poly;
+    size_t enc_len = key->params->pk_len;
+    uint8_t *enc = OPENSSL_malloc(enc_len);
+    WPACKET pkt;
+
+    if (enc == NULL)
+        return 0;
+
+    if (!WPACKET_init_static_len(&pkt, enc, enc_len, 0)
+            || !WPACKET_memcpy(&pkt, key->rho, sizeof(key->rho)))
+        goto err;
+    for (i = 0; i < t1_len; i++)
+        if (!poly_encode_10_bits(t1 + i, &pkt))
+            goto err;
+    if (!WPACKET_get_total_written(&pkt, &written)
+            || written != enc_len)
+        goto err;
+    OPENSSL_free(key->pub_encoding);
+    key->pub_encoding = enc;
+    ret = 1;
+err:
+    WPACKET_finish(&pkt);
+    if (ret == 0)
+        OPENSSL_free(enc);
+    return ret;
+}
+
+/*
+ * @brief The reverse of ossl_ml_dsa_pk_encode().
+ * See FIPS 204, Algorithm 23, pkDecode().
+ *
+ * @param in An encoded public key.
+ * @param in_len The size of |in|
+ * @param key A key object to store the decoded public key into.
+ *
+ * @returns 1 if the public key was decoded successfully or 0 otherwise.
+ */
+int ossl_ml_dsa_pk_decode(ML_DSA_KEY *key, const uint8_t *in, size_t in_len)
+{
+    int ret = 0;
+    size_t i;
+    PACKET pkt;
+    EVP_MD_CTX *ctx;
+
+    if (key->pub_encoding != NULL)
+        return 0; /* Do not allow key mutation */
```

Comment:
Note, this will I believe also cover the case of a private key already being loaded, because loading a private key precomputes a public key.